feature: add backdrop settings dialog with 4-corner color customization

andreasbkrueger opened the pull request:
Includes:
• individual color pickers for each corner
• predefined backdrop presets

andreasbkrueger committed:
feature: add backdrop settings dialog with 4-corner color customization

Includes:
• individual color pickers for each corner
• predefined backdrop presets

andreasbkrueger committed:
fix(render): clear depth buffer after backdrop to prevent it from overlaying 3D model

andreasbkrueger commented:
Hello, this is now my working suggestion on how to customize the background colors - I did not update the version as I'm not sure of the impact this would have - also I guess this is not my place to do so in this commit.

I know that the pre-defined colors are hard coded for now. They could be placed in a config file, for example, where the user itself could add there own custom definitions, but I think this should be another feature and would be to much for this first implementation here.

Hope I did not miss anything or broke something on the way - I was careful, but am not so comfortable with C++ and multi platform apps.

edit:
I'm not much of a designer, but I added an icon for the background settings. Open to any suggestions for improvements! :smile:

wdaniau commented:
Hi Andreas,

I like it very much! Very cool addition.

Some remarks though : 

- There's an unexpected behavior of the "Custom colors" preset in my opinion : 
  - custom colors are correctly saved, as if I close the app and restart it, my custom colors are back, however if I change the preset and come back to custom, my custom colors are erased.
 - please add the B shortcut in the Help menu
 - about the icon, it is quite self-explanatory and there's the tooltip so it's ok. Not very fan of this clear blue, but personal taste ^^.
 - also I'd like to have the original fstl background colors in the available presets, maybe with name "Original" or "Standard"

William

andreasbkrueger committed:
fixes after review (ui): make custom corner color behavior more consistent; warn when preset edits override custom colors

andreasbkrueger committed:
fixes after review (ui): added background settings shortcut to help menu

andreasbkrueger commented:
Hi William,

thank you for the feedback. The colors selection was indeed a bit unclear. It took me a while, but I think it is better now. 

- On first install/start: standard colors
- User changes the preset: colors changes (not the custom colors)
- User clicks on a color picker: system warns, that custom colors are set with new colors (only if the user has a preset selected)
- toggle between custom colors and predefined colors is now possible as expected

The shortcut is in the help menu now.

I tried to change the icon with another gradient, but I'm really bad at this and had to give up - sorry :D - maybe you can give it a shot, or we leave it like this for the moment.

Hope you like the main changes anyway.

Andreas

wdaniau commented:
Nice, 
I did not understood at first how the warning was supposed to be shown to the user because I only changed colors in custom preset ^^ .
When I first make it appear, I thought that one option could have been to greyed out the button when not in custom colors mode. But the way you implemented it makes sense and is better, as the presets can be used as templates, however the message was not clear to me. I suggest changing the warning to a more precise sentence, something like :
"Changing a color in a preset will use the preset colors as a template replacing your current custom colors. Continue?"

Two more things, I would like to have a similar behavior as the preference window : 

- toggle functionality when clicking on the button in the toolbar
- remember window position

And I think we will be good to merge!


What do you think of this for the icon? 
<img width="64" height="64" alt="backdrop-settings_new" src="https://github.com/user-attachments/assets/b8c5eadd-4322-4dca-97f9-bc833f5efca7" />

And in place : 
<img width="880" height="605" alt="Screenshot_20251113_223656" src="https://github.com/user-attachments/assets/d955d5e3-580c-4e53-bbb5-56aa97cf2dd0" />

andreasbkrueger committed:
fixes after review (ui):
- remember backdrop settings dialog position
- toggle behavior for backdrop settings dialog
- custom color override warning more concise now
- changed the backdrop settings icon

andreasbkrueger commented:
Alright - I made the changes.

I did not realized there was a toggle and size/position remembering in the other setting dialog :sweat_smile: thanks for the heads up.

The message makes it more clear now - true!

I used your icon and I think its definitely better now.

To be honest, I tried to add a small border around the background with the corner radius, but was just not able to pull it of. I think the contrast is a bit off on the bottom right now. I am really useless with a graphic tools - that's why we always had designers in the places I worked for. :laughing: - if you can live with the icon as is now, I can too!

wdaniau commented on src/backdropsettingsdialog.cpp:
```diff
@@ -0,0 +1,324 @@
+void BackdropSettingsDialog::resizeEvent(QResizeEvent *event)
+{
+    QSettings().setValue(SETTINGS_DIALOG_GEOMETRY, saveGeometry());
+    QDialog::resizeEvent(event);
```
```suggestion
    QWidget::resizeEvent(event);
```

wdaniau reviewed (approved):
Changed `QDialog::resizeEvent` to `QWidget::resizeEvent` for homogeneity
